src/main/scala/za/co/absa/spark/commons/implicits/StructTypeImplicits.scala

```diff
@@ -0,0 +1,432 @@
+     *
+     * @param field  field to work with
+     * @return       Metadata "sourcecolumn" if it exists or field.name
+     */
+    def getFieldNameOverriddenByMetadata(field: StructField): String = {
+      if (field.metadata.contains(MetadataKeys.SourceColumn)) {
+        field.metadata.getString(MetadataKeys.SourceColumn)
+      } else {
+        field.name
+      }
+    }
+
+    /**
+     * Get paths for all array fields in the schema
+     *
+     * @return Seq of dot separated paths of fields in the schema, which are of type Array
+     */
+    def getAllArrayPaths(): Seq[String] = {
+      schema.fields.flatMap(f => getAllArraySubPaths("", f.name, f.dataType)).toSeq
+    }
+
+    /**
+     * Get a closest unique column name
+     *
+     * @param desiredName A prefix to use for the column name
+     * @param schema      A schema to validate if the column already exists
+     * @return A name that can be used as a unique column name
+     */
+    def getClosestUniqueName(desiredName: String, schema: StructType): String = {
+      var exists = true
+      var columnName = ""
+      var i = 0
+      while (exists) {
+        columnName = if (i == 0) desiredName else s"${desiredName}_$i"
+        exists = schema.fields.exists(_.name.compareToIgnoreCase(columnName) == 0)
+        i += 1
+      }
+      columnName
+    }
+
+    /**
+     * Checks if a field is the only field in a struct
+     *
+     * @param column A column to check
+     * @return true if the column is the only column in a struct
+     */
+    def isOnlyField(column: String): Boolean = {
+      def structHelper(structField: StructType, path: Seq[String]): Boolean = {
+        val currentField = path.head
+        val isLeaf = path.lengthCompare(1) <= 0
+        var isOnlyField = false
+        structField.fields.foreach(field =>
+          if (field.name == currentField) {
+            if (isLeaf) {
+              isOnlyField = structField.fields.length == 1
+            } else {
+              field.dataType match {
+                case st: StructType =>
+                  isOnlyField = structHelper(st, path.tail)
+                case _: ArrayType =>
+                  throw new IllegalArgumentException(
+                    s"SchemaUtils.isOnlyField() does not support checking struct fields inside an array")
+                case _ =>
+                  throw new IllegalArgumentException(
+                    s"Primitive fields cannot have child fields $currentField is a primitive in $column")
+              }
+            }
+          }
+        )
+        isOnlyField
+      }
+      val path = column.split('.')
+      structHelper(schema, path)
+    }
+
+    /**
+     * Checks if a field is an array that is not nested in another array
+     *
+     * @param fieldPathName A field to check
+     * @return true if a field is an array that is not nested in another array
+     */
+    def isNonNestedArray(fieldPathName: String): Boolean = {
```

Comment:
This and `isOnlyField` might need a bit more work, rewrite, but I am not able to complete that now.